MDEV-18842 Unfortunate error message when the same column is used for…

nayuta-yanagisawa opened the pull request:
… application period start and end

An application-time period must be composed of two different temporal columns. We added a check that ensures that the above condition is met.

https://jira.mariadb.org/browse/MDEV-18842

nayuta-yanagisawa commented:
The Unfortunate error mentioned in the JIRA issue seems to be raised by the following code. If two temporal columns composing an application-time period were the same, row_end would be NULL.

```
  bool res= period_info.check_field(row_start, period.start.str)
            || period_info.check_field(row_end, period.end.str);
```

an3l reviewed (commented):
Should target version be 10.4?
Also would it make sense to change this on parser level `sql/sql_yacc`?
```
period_for_application_time:
```

nayuta-yanagisawa commented:
> Should target version be 10.4?

[Getting started for developers](https://mariadb.org/get-involved/getting-started-for-developers/) says as the following. 

> Development of new features is done on the development master branch, currently 10.6. In general all development is done on the master branch first, and later high-priority bug fixes might be backported to stable release branches (10.5, 10.4, 10.3, 10.2, 10.1).

Thus, I decided to target 10.6 because fixing a bug is a sort of "development". However, I'm not so sure that this is the correct way because I'm quite new to MariaDB development. Should I change the base branch to 10.4?

> Also would it would make sense to change this on parser level sql/sql_yacc?

I think this makes sense. I will explore that direction.

nayuta-yanagisawa commented:
Force pushed for rerunning CIs.

FooBarrior commented on sql/share/errmsg-utf8.txt:
```diff
@@ -7962,3 +7962,5 @@ ER_NOT_ALLOWED_IN_THIS_CONTEXT
         eng "'%-.128s' is not allowed in this context"
 ER_DATA_WAS_COMMITED_UNDER_ROLLBACK
         eng "Engine %s does not support rollback. Changes were committed during rollback call"
+ER_PERIOD_PAIR_OF_SAME_COLUMNS
```
We can't add new error codes in 10.4 and earlier. Better use some existing error message, like `ER_FIELD_SPECIFIED_TWICE`

nayuta-yanagisawa replied:
@FooBarrior Thank you for your comment. I've fixed along with your advice.

FooBarrior reviewed (changes requested):
Nayuta, please fix the issue I've addressed and ping me again

nayuta-yanagisawa commented:
Sorry for bothering you but, honestly, I'm a bit confused about various lex strings and their relation. I'm not sure that it is appropriate to add `streq` to `Lex_ident_sys_st` while there is `Lex_ident` with a comparison function.

nayuta-yanagisawa commented:
By the way, should I change the base branch to 10.4?

FooBarrior commented:
@nayuta-yanagisawa Yes, change the branch to 10.4, pls.

Mariadb currently supports versions from 10.2 and newer, this means that the bugs there are to be fixed.
Yes, bugfixing is a development, but developing a new feature is different. Mostly the new feature tickets have a "Task" type, not "Bug".

Though we can do slight refactorings in older code, if it is required to fix a bug.

FooBarrior commented:
@nayuta-yanagisawa you should use `lex_string_cmp`, not strcmp, to compare identifiers. Like this:
```
if (lex_string_cmp(system_charset_info, &kp->field_name,
                                  &f->field_name) != 0)
```

IMO it's fine to add `streq` there, but let's see what @an3l thinks on it

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MariaDB/server?pullRequest=1682) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/MariaDB/server?pullRequest=1682) before we can accept your contribution.<br/>**3** out of **18** committers have signed the CLA.<br/><br/>:white_check_mark: cvicentiu<br/>:white_check_mark: grooverdan<br/>:white_check_mark: nayuta-yanagisawa<br/>:x: dr-m<br/>:x: vaintroub<br/>:x: sanja-byelkin<br/>:x: mkaruza<br/>:x: Thirunarayanan<br/>:x: sujatha-s<br/>:x: montywi<br/>:x: janlindstrom<br/>:x: sciascid<br/>:x: dbart<br/>:x: igorbabaev<br/>:x: vuvova<br/>:x: spetrunia<br/>:x: FooBarrior<br/>:x: midenok<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MariaDB/server?pullRequest=1682) it.</sub>

nayuta-yanagisawa commented:
@FooBarrior Thank you very much for your kind instruction.  I've changed the base branch to 10.4.

Also, I've modified the code to use `lex_string_cmp` rather than `strcmp`.

nayuta-yanagisawa commented:
When I changed the base branch many commits temporarily mixed in this pull request. That seems to cause an unintentional review request to @ottok . Sorry for bothering you.

an3l commented:
@FooBarrior I'm ok with suggested function with comparing the strings. However I wouldn't create additional member for one-time compare. In fact I wouldn't use it at all:
@nayuta-yanagisawa I was thinking to do something like this instead, if @FooBarrior  is ok with:
```
--- a/sql/sql_yacc.yy
+++ b/sql/sql_yacc.yy
@@ -6644,6 +6644,11 @@ period_for_system_time:
 period_for_application_time:
           FOR_SYM ident '(' ident ',' ident ')'
           {
+            if (strcmp($4.str, $6.str) == 0)
+            {
+                my_error(ER_FIELD_SPECIFIED_TWICE, MYF(0), $4.str);
+                return TRUE;
+            }
             if (Lex->add_period($2, $4, $6))
               MYSQL_YYABORT;
           }

```

nayuta-yanagisawa commented:
I agree that defining a member function for a single-use is probably redundant. So, I will remove it. However, I think it is not so bad idea to encapsulate checks into `add_period` function. Thus, I would like to keep  ER_FIELD_SPECIFIED_TWICE check in `add_period` function.

If reviewers are OK, I will fix the code along with the above explained way.

FooBarrior commented:
I think it's better to leave the check in add_period(). `sql_yacc.yy` is already giantic, so better to tend to store the grammar there only, and as less logic as possible

nayuta-yanagisawa commented:
I've fixed the code along with the [above-mentioned way](https://github.com/MariaDB/server/pull/1682#issuecomment-714194869).

an3l commented:
@FooBarrior^ seems ok according to me.
